### DISCUSSION TITLE
linked_clone?

### DISCUSSION THREAD

**Discussion by Jeansen** [vagrant-libvirt/vagrant-libvirt#1669] [Q&A]
Maybe I am missing something, but I could not find any option that would be equal to linked_clone provided by the virtualbox provider. Am I missing something?

**Comment by electrofelix** [comment]
Backing files would be the equivalent to Virtualbox's linked clone, and this is what vagrant-libvirt uses for all machines.

What we don't have is the ability to run without backing files, where the image uploaded into the libvirt storage is used directly. See https://github.com/vagrant-libvirt/vagrant-libvirt/blob/4fe53477b204fbb9626d47905c328808b8d2a742/lib/vagrant-libvirt/action/create_domain_volume.rb#L64 where every domain volume is created using the image uploaded from the box as a backing file.

Possibly where people need maximum performance using a direct image would be a little faster.

**Reply by Jeansen** [reply]
@electrofelix Thanks for the quick and detailed answer. I must admit, my Ruby is very limited and with respect to the linked_clone setting I thought there would be something simpler. In my use case, I'd like to first spin up one VM, do some provisioning and then use that provisioned vm/image as a template (linked clone) for all other machines to come (which then only have minor changes).

**Reply by electrofelix** [reply]
So I may not have understood how `linked_clone` for VirtualBox works in vagrant correctly. I understood it to mean that while another VM would appear in VirtualBox, you would not be able to modify that VM through vagrant, and behaviour is undefined if you do modify it since the `linked_clone` machines are using copy-on-write snapshots on top, which if something changes underneath in the base image it can corrupt the result of trying to run the VMs with differences applied on top.

Typically this means that if you want to use a VM that has been modified as the base, it involves use of `vagrant package` to convert the VM to a box that is then used by your subsequent Vagrantfile machines.

Due to vagrant-libvirt's requirement for a management network (working on removing that but it's a while away), calling package and having a working box afterwards can be a little more complex than it needs to be depending on the distro and it's behaviour.

I do have some documentation that I've yet to include that will be added to https://vagrant-libvirt.github.io/vagrant-libvirt/packaging.html in the near future, but in the mean time the best guide is the acceptance test
https://github.com/vagrant-libvirt/vagrant-libvirt/blob/980db1049a2752ea41f56d5134282e9f28fd5f80/spec/acceptance/package_domain_spec.rb#L56-L89

The vagrantfile used for the base machine is https://github.com/vagrant-libvirt/vagrant-libvirt/blob/980db1049a2752ea41f56d5134282e9f28fd5f80/spec/acceptance/support-skeletons/package_complex/Vagrantfile, note the suggestion to disable inserting the replacement ssh key

Additionally it makes use of a script to perform some clean up on the final image extracted https://github.com/vagrant-libvirt/vagrant-libvirt/blob/main/spec/acceptance/support-skeletons/package_complex/scripts/sysprep.sh

While it's possible to perform some of the clean up steps as part of your final steps of provisioning, that can make it a bit more painful to bring up that machine to make further provisioning changes without having to start from scratch each time.


With all that said, it does sound like an interesting idea to consider a VM that is provisioned, halted and then used as the base for others directly without being packaged. The problem with this is usually the cloned machines will end up getting the same network address due to having the same machine id which is difficult to replace before it has connected to the network. So until I've removed the requirement to have a network running in order to provision and access vagrant-libvirt machines, limited to needing to use the package and add box approach.

**Comment by Jeansen** [comment]
@electrofelix As always your answers are much appreciated with that amount of detail. I think I'll wait until you have removed the network requirement. I'd be happy to test anything, If you want me to.
